core/sched.c

```diff
@@ -181,6 +190,12 @@ void sched_switch(uint16_t other_prio)
 
 NORETURN void sched_task_exit(void)
 {
+    /* Prevent optimization of these symbols */
+    max_threads = max_threads;
```

Comment:
@gebart

Doesn't work. It is caused by the linker garbage collection:
```bash
export CFLAGS_LINK  = -ffunction-sections -fdata-sections
export LINKFLAGS += -Wl,--gc-sections
```
So there doesn't seem to be a way to prevent optimization from within the compiler other than removing `-fdata-sections`. But that's no option.

Adding
```
export LINKFLAGS += -Wl,-u,max_threads
```
to `boards/Makefile.include.cortexm_common` keeps the variable from being optimized away, but that's not a nice solution either.

Another possible solution would be to define a new section in RAM (linker script) with `KEEP(this-section)` and moving the variable into that section.

I would rather like to have a solution where it's only marked next to variable in source, but there doesn't seem to be a way. The latter solution is not far away from that at least.